openedx/openedx-platform
# Store split modulestore's course indexes in Django/MySQL

## Description

This change modifies spit modulestore so that split's course index data is read from and written to MySQL instead of MongoDB.

Background: Split modulestore persists data in three MongoDB "collections": `course_index` (list of courses and the current version of each), `structure` (outline of the courses, and some XBlock fields), and `definition` (other XBlock fields). While "structure" and "definition" data can get very large, which is one of the reasons MongoDB was chosen for modulestore, the course index data is very small.

By moving course index data to MySQL / a django model, we get these advantages:

* Full history of changes to the course index data is preserved (using `django-simple-history`)
* It's much easier to inspect the list of courses and libraries from the django admin:
    ![Screen Shot 2021-05-07 at 3 18 59 PM](https://user-images.githubusercontent.com/945577/117514218-382e5a00-af48-11eb-8259-7a2698f06a83.png)
* It's much easier to "reset" a corrupted course to a known working state, by using the simple-history revert tools from the django admin.
* The remaining MongoDB collections (`structure` and `definition`) are essentially just used as key-value stores of large JSON data structures.
    - This PR paves the way for a future changes that allows migrating courses one at a time from MongoDB to S3, and thus eliminating any use of MongoDB by split modulestore, simplifying the stack.

## Deadline

None

## Performance

It is yet unknown whether this approach is faster or slower. Writes are expected to be slower at first due to writing through to both MySQL and MongoDB (see below).

## Migration

This PR is written for immediate cutover: it includes a data migration that copies all course indexes into MySQL, and it starts doing all reads exclusively from MySQL. However, writes are persisted to both MySQL and MongoDB so that this PR can be reverted without any loss of data.

## Testing instructions

Create, edit, and learn from both courses and content libraries using Studio. Ensure draft+preview+publish works.

Try rolling back to a version before this PR (that uses Mongo only) and verify that the new updates you made to the course are still there. (This tests that the double-write to MySQL+MongoDB is working, to make this safer to deploy and easier to roll back.)
